File: crates/ferrogate-cli/src/network_access.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-06
// description: Pre-authentication IP/CIDR allowlisting and per-source-IP
// flood throttling, applied ahead of virtual-key resolution (issue #166).

use std::collections::HashMap;
use std::net::IpAddr;
use std::sync::Mutex;

use http::HeaderMap;

/// A parsed IPv4/IPv6 CIDR range. A bare IP address (no `/n` suffix) is
/// treated as an exact match (`/32` for IPv4, `/128` for IPv6).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct IpCidr {
    network: IpAddr,
    prefix_len: u8,
}

impl IpCidr {
    pub(crate) fn parse(raw: &str) -> Result<Self, String> {
        let trimmed = raw.trim();
        if trimmed.is_empty() {
            return Err("empty CIDR/IP entry".to_string());
        }
        let (ip_part, prefix_part) = match trimmed.split_once('/') {
            Some((ip, prefix)) => (ip, Some(prefix)),
            None => (trimmed, None),
        };
        let network: IpAddr = ip_part
            .parse()
            .map_err(|_| format!("invalid IP address: {ip_part}"))?;
        let max_prefix: u8 = match network {
            IpAddr::V4(_) => 32,
            IpAddr::V6(_) => 128,
        };
        let prefix_len = match prefix_part {
            Some(raw_prefix) => raw_prefix
                .parse::<u8>()
                .map_err(|_| format!("invalid prefix length: {raw_prefix}"))?,
            None => max_prefix,
        };
        if prefix_len > max_prefix {
            return Err(format!(
                "prefix length /{prefix_len} exceeds maximum /{max_prefix} for {network}"
            ));
        }
        Ok(Self {
            network,
            prefix_len,
        })
    }

    pub(crate) fn contains(&self, candidate: &IpAddr) -> bool {
        match (self.network, candidate) {
            (IpAddr::V4(network), IpAddr::V4(candidate)) => {
                masked_eq_u32(u32::from(network), u32::from(*candidate), self.prefix_len)
            }
            (IpAddr::V6(network), IpAddr::V6(candidate)) => {
                masked_eq_u128(u128::from(network), u128::from(*candidate), self.prefix_len)
            }
            _ => false,
        }
    }
}

/// Compares the top `prefix_len` bits of two 32-bit addresses. A
/// `prefix_len` of 0 matches everything; 32 requires an exact match.
fn masked_eq_u32(a: u32, b: u32, prefix_len: u8) -> bool {
    let shift = 32u32.saturating_sub(u32::from(prefix_len));
    let mask = u32::MAX.checked_shl(shift).unwrap_or(0);
    (a & mask) == (b & mask)
}

/// Same as [`masked_eq_u32`] but for 128-bit (IPv6) addresses.
fn masked_eq_u128(a: u128, b: u128, prefix_len: u8) -> bool {
    let shift = 128u32.saturating_sub(u32::from(prefix_len));
    let mask = u128::MAX.checked_shl(shift).unwrap_or(0);
    (a & mask) == (b & mask)
}

/// Resolves the client IP used for allowlist/rate-limit decisions: either
/// the raw TCP peer address, or (only when explicitly trusted via config) the
/// entry `trusted_proxy_hops` positions from the RIGHT of `X-Forwarded-For`,
/// falling back to `X-Real-IP`.
///
/// Trusting forwarded headers by default would let any client spoof its
/// apparent source IP, so this only consults them when `trust_forwarded_for`
/// is set (the operator has trusted reverse proxies in front and the gateway is
/// only reachable through them).
///
/// Selection is from the right, not the left: the common load balancers
/// (nginx `$proxy_add_x_forwarded_for`, AWS ALB, GCP LB, Cloudflare) *append*
/// the peer they received from, so with `H` trusted proxies in front the
/// rightmost `H` entries are proxy-authored and the `H`-th-from-right entry is
/// the real client. Everything to its left is client-supplied and therefore
/// spoofable — taking the leftmost entry (the pre-fix behavior) let a client
/// behind an appending proxy forge its apparent source IP, bypassing the
/// pre-auth allowlist and minting a fresh rate-limit window per fake IP.
/// `trusted_proxy_hops` is clamped to at least 1 whenever forwarded headers are
/// trusted.
pub(crate) fn resolve_client_ip(
    headers: &HeaderMap,
    peer_addr: Option<IpAddr>,
    trust_forwarded_for: bool,
    trusted_proxy_hops: u32,
) -> Option<IpAddr> {
    if trust_forwarded_for {
        let hops = (trusted_proxy_hops.max(1)) as usize;
        if let Some(forwarded) = headers
            .get("x-forwarded-for")
            .and_then(|value| value.to_str().ok())
        {
            let entries: Vec<&str> = forwarded
                .split(',')
                .map(str::trim)
                .filter(|entry| !entry.is_empty())
                .collect();
            // The client is `hops` entries from the right. If the chain is
            // shorter than the configured hop count (fewer proxies than
            // expected — a misconfiguration or a direct connection that skipped
            // the proxy fleet), do NOT fall back to a client-supplied entry;
            // drop to X-Real-IP / peer_addr below, which fail closed.
            if let Some(entry) = entries.len().checked_sub(hops).and_then(|i| entries.get(i)) {
                if let Ok(ip) = entry.parse::<IpAddr>() {
                    return Some(ip);
                }
            }
        }
        if let Some(real_ip) = headers
            .get("x-real-ip")
            .and_then(|value| value.to_str().ok())
            .and_then(|value| value.trim().parse::<IpAddr>().ok())
        {
            return Some(real_ip);
        }
    }
    peer_addr
}

/// A pragmatic bound on how many distinct source IPs the pre-auth limiter
/// tracks at once. Once exceeded, the tracking map is cleared rather than
/// grown further, so a spoofed-source flood cannot turn the limiter itself
/// into a memory-exhaustion vector. This trades a brief window of
/// under-counting (right after a clear) for a hard memory ceiling.
const MAX_TRACKED_SOURCE_IPS: usize = 100_000;

/// Fixed-window (per-minute), per-source-IP request counter used to throttle
/// floods before any virtual-key/storage lookup happens. Deliberately
/// simpler than the per-key `ClusterCounterBackend` machinery: this is a
/// coarse, single-node, in-process flood gate, not a billing-accurate quota.
#[derive(Debug, Default)]
pub(crate) struct UnauthenticatedIpRateLimiter {
    windows: Mutex<HashMap<IpAddr, (u64, u64)>>,
}

impl UnauthenticatedIpRateLimiter {
    /// Returns `true` if `ip` is still within `limit` requests for the
    /// current `now_minute` window (a monotonically increasing minute
    /// counter, e.g. `unix_seconds / 60`).
    pub(crate) fn allow(&self, ip: IpAddr, now_minute: u64, limit: u64) -> bool {
        let mut windows = self
            .windows
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        if windows.len() >= MAX_TRACKED_SOURCE_IPS && !windows.contains_key(&ip) {
            windows.clear();
        }
        let entry = windows.entry(ip).or_insert((now_minute, 0));
        if entry.0 != now_minute {
            *entry = (now_minute, 0);
        }
        entry.1 += 1;
        entry.1 <= limit
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_bare_ipv4_as_exact_match() {
        let cidr = IpCidr::parse("203.0.113.7").unwrap();
        assert!(cidr.contains(&"203.0.113.7".parse().unwrap()));
        assert!(!cidr.contains(&"203.0.113.8".parse().unwrap()));
    }

    #[test]
    fn parses_ipv4_cidr_range() {
        let cidr = IpCidr::parse("10.0.0.0/8").unwrap();
        assert!(cidr.contains(&"10.1.2.3".parse().unwrap()));
        assert!(!cidr.contains(&"11.0.0.1".parse().unwrap()));
    }

    #[test]
    fn parses_ipv4_slash_zero_as_match_all() {
        let cidr = IpCidr::parse("0.0.0.0/0").unwrap();
        assert!(cidr.contains(&"1.2.3.4".parse().unwrap()));
        assert!(cidr.contains(&"255.255.255.255".parse().unwrap()));
    }

    #[test]
    fn parses_ipv4_slash_32_as_exact_match() {
        let cidr = IpCidr::parse("203.0.113.7/32").unwrap();
        assert!(cidr.contains(&"203.0.113.7".parse().unwrap()));
        assert!(!cidr.contains(&"203.0.113.6".parse().unwrap()));
    }

    #[test]
    fn parses_ipv6_cidr_range() {
        let cidr = IpCidr::parse("2001:db8::/32").unwrap();
        assert!(cidr.contains(&"2001:db8::1".parse().unwrap()));
        assert!(!cidr.contains(&"2001:db9::1".parse().unwrap()));
    }

    #[test]
    fn does_not_match_across_address_families() {
        let cidr = IpCidr::parse("10.0.0.0/8").unwrap();
        assert!(!cidr.contains(&"::1".parse().unwrap()));
    }

    #[test]
    fn rejects_invalid_ip() {
        assert!(IpCidr::parse("not-an-ip").is_err());
    }

    #[test]
    fn rejects_invalid_prefix_length() {
        assert!(IpCidr::parse("10.0.0.0/33").is_err());
        assert!(IpCidr::parse("10.0.0.0/abc").is_err());
    }

    #[test]
    fn rejects_empty_entry() {
        assert!(IpCidr::parse("").is_err());
        assert!(IpCidr::parse("   ").is_err());
    }

    fn headers_with(pairs: &[(&str, &str)]) -> HeaderMap {
        let mut headers = HeaderMap::new();
        for (name, value) in pairs {
            headers.insert(
                http::HeaderName::from_bytes(name.as_bytes()).unwrap(),
                http::HeaderValue::from_str(value).unwrap(),
            );
        }
        headers
    }

    #[test]
    fn ignores_forwarded_headers_when_not_trusted() {
        let headers = headers_with(&[("x-forwarded-for", "198.51.100.9")]);
        let peer: IpAddr = "127.0.0.1".parse().unwrap();
        assert_eq!(
            resolve_client_ip(&headers, Some(peer), false, 1),
            Some(peer)
        );
    }

    #[test]
    fn uses_rightmost_untrusted_forwarded_for_entry_with_single_hop() {
        // Behind one appending proxy the chain is `<spoofable client value>,
        // <real client the proxy appended>`. With one trusted hop the real
        // client is the rightmost entry.
        let headers = headers_with(&[("x-forwarded-for", "203.0.113.7, 198.51.100.9")]);
        let peer: IpAddr = "127.0.0.1".parse().unwrap();
        assert_eq!(
            resolve_client_ip(&headers, Some(peer), true, 1),
            Some("198.51.100.9".parse().unwrap())
        );
    }

    #[test]
    fn leftmost_forwarded_for_entry_cannot_spoof_an_allowlisted_source() {
        // The exploit: an attacker sends `X-Forwarded-For: 10.0.0.1` (an
        // allowlisted internal IP); the appending proxy yields
        // `10.0.0.1, <attacker_ip>`. The old leftmost selection returned the
        // spoofed 10.0.0.1; the fix returns the attacker's real appended IP.
        let headers = headers_with(&[("x-forwarded-for", "10.0.0.1, 203.0.113.66")]);
        let peer: IpAddr = "127.0.0.1".parse().unwrap();
        let resolved = resolve_client_ip(&headers, Some(peer), true, 1).unwrap();
        assert_eq!(resolved, "203.0.113.66".parse::<IpAddr>().unwrap());
        assert_ne!(resolved, "10.0.0.1".parse::<IpAddr>().unwrap());
    }

    #[test]
    fn selects_client_by_hop_count_through_a_proxy_chain() {
        // Two trusted proxies: `<spoofable>, <client>, <proxy1>`. With two hops
        // the client is two entries from the right.
        let headers = headers_with(&[("x-forwarded-for", "203.0.113.7, 198.51.100.9, 192.0.2.10")]);
        let peer: IpAddr = "127.0.0.1".parse().unwrap();
        assert_eq!(
            resolve_client_ip(&headers, Some(peer), true, 2),
            Some("198.51.100.9".parse().unwrap())
        );
    }

    #[test]
    fn falls_back_when_chain_is_shorter_than_the_hop_count() {
        // Fewer entries than configured hops (a direct connection that skipped
        // the proxy fleet): do NOT trust a client-supplied entry -- fall back to
        // X-Real-IP / peer.
        let headers = headers_with(&[
            ("x-forwarded-for", "203.0.113.7"),
            ("x-real-ip", "198.51.100.9"),
        ]);
        let peer: IpAddr = "127.0.0.1".parse().unwrap();
        assert_eq!(
            resolve_client_ip(&headers, Some(peer), true, 2),
            Some("198.51.100.9".parse().unwrap())
        );
        let peer_only = headers_with(&[("x-forwarded-for", "203.0.113.7")]);
        assert_eq!(
            resolve_client_ip(&peer_only, Some(peer), true, 2),
            Some(peer)
        );
    }

    #[test]
    fn falls_back_to_real_ip_header_when_trusted() {
        let headers = headers_with(&[("x-real-ip", "198.51.100.9")]);
        let peer: IpAddr = "127.0.0.1".parse().unwrap();
        assert_eq!(
            resolve_client_ip(&headers, Some(peer), true, 1),
            Some("198.51.100.9".parse().unwrap())
        );
    }

    #[test]
    fn falls_back_to_peer_addr_when_no_forwarded_headers_present() {
        let headers = HeaderMap::new();
        let peer: IpAddr = "127.0.0.1".parse().unwrap();
        assert_eq!(resolve_client_ip(&headers, Some(peer), true, 1), Some(peer));
    }

    #[test]
    fn rate_limiter_allows_up_to_limit_then_denies() {
        let limiter = UnauthenticatedIpRateLimiter::default();
        let ip: IpAddr = "203.0.113.1".parse().unwrap();
        for _ in 0..5 {
            assert!(limiter.allow(ip, 100, 5));
        }
        assert!(!limiter.allow(ip, 100, 5));
    }

    #[test]
    fn rate_limiter_resets_on_new_window() {
        let limiter = UnauthenticatedIpRateLimiter::default();
        let ip: IpAddr = "203.0.113.1".parse().unwrap();
        for _ in 0..5 {
            assert!(limiter.allow(ip, 100, 5));
        }
        assert!(!limiter.allow(ip, 100, 5));
        assert!(limiter.allow(ip, 101, 5));
    }

    #[test]
    fn rate_limiter_tracks_sources_independently() {
        let limiter = UnauthenticatedIpRateLimiter::default();
        let a: IpAddr = "203.0.113.1".parse().unwrap();
        let b: IpAddr = "203.0.113.2".parse().unwrap();
        for _ in 0..5 {
            assert!(limiter.allow(a, 100, 5));
        }
        assert!(!limiter.allow(a, 100, 5));
        assert!(limiter.allow(b, 100, 5));
    }
}

File: packages/config/src/network-access.ts
/**
 * Port of `ferrogate-config`'s `config/network_access.rs` (inventory §5.4,
 * "Network access", issue #166): pre-authentication IP/CIDR allowlisting,
 * client-IP resolution (anti-spoof `X-Forwarded-For`-from-the-right), and the
 * per-source-IP flood limiter.
 *
 * PORT-TODO(inventory §5.8) — PLATFORM LIMIT, NOT CLOSED (two distinct ones).
 *
 * 1. NO PEER ADDRESS. The Rust gate started from the TCP peer address it read
 *    off the pingora socket. A Worker never sees a socket: workerd hands the
 *    handler a `Request` and nothing else, so there is no `peer_addr` to read.
 *    CLOSEST BEHAVIOR IMPLEMENTED: `resolveClientIp` keeps the Rust signature
 *    and semantics exactly (including the anti-spoof XFF-from-the-RIGHT hop
 *    selection and the fail-closed short-chain path); the CALLER substitutes
 *    Cloudflare's `CF-Connecting-IP` header — which the edge sets and strips
 *    from client input — for `peerAddr`. That header is the only value on this
 *    platform with the same trust property as the Rust peer address.
 *
 * 2. NO SHARED MUTABLE STATE. `UnauthenticatedIpRateLimiter`'s Map is
 *    ISOLATE-LOCAL and always will be: Workers isolates share no memory, are
 *    created/evicted per colo without coordination, and the only cross-isolate
 *    counters (Durable Objects, the Rate Limiting binding) are ASYNC, so they
 *    cannot be hidden behind this synchronous `allow()`.
 *    CLOSEST BEHAVIOR IMPLEMENTED: the Rust fixed-window arithmetic and the
 *    {@link MAX_TRACKED_SOURCE_IPS} memory-exhaustion guard are ported verbatim
 *    and kept synchronous, so a Durable Object can host ONE instance of this
 *    class and get the exact Rust decision; used directly in a Worker it
 *    throttles per isolate, which is a weaker bound, not a different rule.
 *
 * The CIDR matcher itself (`IpCidr`, v4/v6, no cross-family match) is a pure
 * port with no platform caveat. Pinned by `platform-limits.test.ts`.
 */

/** A parsed IPv4/IPv6 CIDR. A bare IP (no `/n`) is an exact match (/32 or /128). */
export class IpCidr {
  private readonly network: bigint;
  private readonly prefixLen: number;
  private readonly isV6: boolean;

  private constructor(network: bigint, prefixLen: number, isV6: boolean) {
    this.network = network;
    this.prefixLen = prefixLen;
    this.isV6 = isV6;
  }

  /** Parse a CIDR/IP entry. Throws (with the Rust message) on any malformed input. */
  static parse(raw: string): IpCidr {
    const trimmed = raw.trim();
    if (trimmed.length === 0) throw new Error("empty CIDR/IP entry");
    const slash = trimmed.indexOf("/");
    const ipPart = slash === -1 ? trimmed : trimmed.slice(0, slash);
    const prefixPart = slash === -1 ? null : trimmed.slice(slash + 1);

    const parsed = parseIp(ipPart);
    if (parsed === null) throw new Error(`invalid IP address: ${ipPart}`);
    const maxPrefix = parsed.isV6 ? 128 : 32;

    let prefixLen: number;
    if (prefixPart === null) {
      prefixLen = maxPrefix;
    } else {
      if (!/^\d+$/.test(prefixPart)) throw new Error(`invalid prefix length: ${prefixPart}`);
      prefixLen = Number.parseInt(prefixPart, 10);
    }
    if (prefixLen > maxPrefix) {
      throw new Error(
        `prefix length /${prefixLen} exceeds maximum /${maxPrefix} for ${ipPart}`,
      );
    }
    return new IpCidr(parsed.value, prefixLen, parsed.isV6);
  }

  /** Whether `candidate` (a raw IP string) falls inside this range. */
  contains(candidate: string): boolean {
    const parsed = parseIp(candidate);
    if (parsed === null) return false;
    if (parsed.isV6 !== this.isV6) return false; // no cross-family match
    const bits = this.isV6 ? 128 : 32;
    const shift = BigInt(bits - this.prefixLen);
    const full = this.isV6 ? (1n << 128n) - 1n : (1n << 32n) - 1n;
    const mask = shift >= BigInt(bits) ? 0n : (full << shift) & full;
    return (this.network & mask) === (parsed.value & mask);
  }
}

/** `str::parse::<IpAddr>()` — is this string a bare IPv4/IPv6 literal? */
export function isIpAddress(raw: string): boolean {
  return parseIp(raw) !== null;
}

/** Parse a v4 or v6 address into a numeric value, or `null` if malformed. */
function parseIp(raw: string): { value: bigint; isV6: boolean } | null {
  const s = raw.trim();
  if (s.includes(":")) {
    const value = parseIpv6(s);
    return value === null ? null : { value, isV6: true };
  }
  const value = parseIpv4(s);
  return value === null ? null : { value, isV6: false };
}

/**
 * Rust `Ipv4Addr::from_str`. Four decimal octets, each 1–3 digits and <= 255,
 * with **no leading zeros**: std refuses `010.0.0.1` deliberately, because an
 * octal-looking octet is read as 10 by one parser and 8 by another, and that
 * disagreement is the classic allowlist-bypass / SSRF primitive. This is the
 * parser behind BOTH {@link IpCidr} (the pre-auth allowlist) and
 * {@link resolveClientIp}'s `X-Forwarded-For` / `X-Real-IP` scan, so accepting
 * a spelling Rust rejects would mean an attacker-supplied `010.0.0.1` is taken
 * as a client identity here and discarded there.
 */
function parseIpv4(s: string): bigint | null {
  const parts = s.split(".");
  if (parts.length !== 4) return null;
  let value = 0n;
  for (const part of parts) {
    if (!/^\d{1,3}$/.test(part)) return null;
    if (part.length > 1 && part.startsWith("0")) return null;
    const octet = Number.parseInt(part, 10);
    if (octet > 255) return null;
    value = (value << 8n) | BigInt(octet);
  }
  return value;
}

function parseIpv6(s: string): bigint | null {
  // Reject a scope id / anything but hex groups and a single `::`.
  const doubleColon = s.indexOf("::");
  if (doubleColon !== s.lastIndexOf("::")) return null;

  let head: string[];
  let tail: string[];
  if (doubleColon === -1) {
    head = s.split(":");
    tail = [];
  } else {
    head = s.slice(0, doubleColon).split(":").filter((g) => g !== "");
    tail = s.slice(doubleColon + 2).split(":").filter((g) => g !== "");
  }

  // Expand a trailing embedded IPv4 (`::ffff:1.2.3.4`) into two hex groups.
  const expandV4 = (groups: string[]): string[] | null => {
    if (groups.length === 0) return groups;
    const last = groups[groups.length - 1]!;
    if (last.includes(".")) {
      const v4 = parseIpv4(last);
      if (v4 === null) return null;
      const hi = (v4 >> 16n) & 0xffffn;
      const lo = v4 & 0xffffn;
      return [...groups.slice(0, -1), hi.toString(16), lo.toString(16)];
    }
    return groups;
  };
  const expandedHead = expandV4(head);
  const expandedTail = expandV4(tail);
  if (expandedHead === null || expandedTail === null) return null;
  head = expandedHead;
  tail = expandedTail;

  const groupCount = head.length + tail.length;
  if (doubleColon === -1) {
    if (groupCount !== 8) return null;
  } else if (groupCount > 7) {
    return null;
  }
  const zeros = 8 - groupCount;
  const groups = [...head, ...Array(zeros).fill("0"), ...tail];
  if (groups.length !== 8) return null;

  let value = 0n;
  for (const group of groups) {
    if (!/^[0-9a-fA-F]{1,4}$/.test(group)) return null;
    value = (value << 16n) | BigInt(Number.parseInt(group, 16));
  }
  return value;
}

/** A header source: fetch `Headers`, or a plain (any-case) record. */
export type HeaderSource = Headers | Record<string, string | undefined>;

function headerValue(headers: HeaderSource, name: string): string | undefined {
  if (typeof (headers as Headers).get === "function") {
    return (headers as Headers).get(name) ?? undefined;
  }
  const lower = name.toLowerCase();
  for (const [key, value] of Object.entries(headers as Record<string, string | undefined>)) {
    if (key.toLowerCase() === lower) return value ?? undefined;
  }
  return undefined;
}

/**
 * Resolves the client IP for allowlist/rate-limit decisions: the raw peer
 * address, or (only when explicitly trusted) the entry `trustedProxyHops`
 * positions from the RIGHT of `X-Forwarded-For`, falling back to `X-Real-IP`.
 *
 * Selection is from the right, not the left (appending proxies), and it fails
 * closed when the chain is shorter than the hop count — a spoofed leftmost
 * entry can never bypass the allowlist. `trustedProxyHops` is clamped to >= 1.
 */
export function resolveClientIp(
  headers: HeaderSource,
  peerAddr: string | null,
  trustForwardedFor: boolean,
  trustedProxyHops: number,
): string | null {
  if (trustForwardedFor) {
    const hops = Math.max(1, trustedProxyHops);
    const forwarded = headerValue(headers, "x-forwarded-for");
    if (forwarded !== undefined) {
      const entries = forwarded
        .split(",")
        .map((e) => e.trim())
        .filter((e) => e.length > 0);
      const idx = entries.length - hops;
      if (idx >= 0 && idx < entries.length) {
        const entry = entries[idx]!;
        if (parseIp(entry) !== null) return entry;
      }
    }
    const realIp = headerValue(headers, "x-real-ip");
    if (realIp !== undefined) {
      const trimmed = realIp.trim();
      if (parseIp(trimmed) !== null) return trimmed;
    }
  }
  return peerAddr;
}

/**
 * Pragmatic bound on how many distinct source IPs the pre-auth limiter tracks.
 * Once exceeded, the map is cleared rather than grown further, so a spoofed
 * flood cannot turn the limiter into a memory-exhaustion vector.
 */
export const MAX_TRACKED_SOURCE_IPS = 100_000;

/**
 * Fixed-window (per-minute), per-source-IP request counter used to throttle
 * floods before any virtual-key/storage lookup. A coarse single-node flood
 * gate, not a billing-accurate quota.
 */
export class UnauthenticatedIpRateLimiter {
  private readonly windows = new Map<string, [number, number]>();

  /**
   * `true` if `ip` is still within `limit` requests for the current
   * `nowMinute` window (a monotonically increasing minute counter, e.g.
   * `unixSeconds / 60`).
   */
  allow(ip: string, nowMinute: number, limit: number): boolean {
    if (this.windows.size >= MAX_TRACKED_SOURCE_IPS && !this.windows.has(ip)) {
      this.windows.clear();
    }
    let entry = this.windows.get(ip);
    if (entry === undefined || entry[0] !== nowMinute) {
      entry = [nowMinute, 0];
      this.windows.set(ip, entry);
    }
    entry[1] += 1;
    return entry[1] <= limit;
  }
}
